### ISSUE TITLE
this.body setter does not respect this.respond

### ISSUE THREAD

**Issue by peteyy** [koajs/koa#843]
If I explicitly set this.respond to false in my handler like this:

```javascript
app.use(function * () {
  this.respond = false
  this.res.end()
  this.body = 'foo'
})
```

If anything sets the body after the response was ended, I get the following error:
> Error: Can't remove headers after they are sent.

https://github.com/koajs/koa/blob/master/lib/response.js#L174 the error maybe on this line.

**Comment by jonathanong** [comment]
it's like you're just trying to break it...

**Comment by peteyy** [comment]
nope, I just simplified the problem that I've had in a little bit more complex application.

**Comment by PlasmaPower** [comment]
I'm assuming this is happening in your code architecture because sometimes you are setting `this.respond` and sometimes you aren't? Is that the same place that you are also ending the request manually? If this is the case, why are you calling next after that? In fact, why conditionally use `this.respond` at all? It seems to me like this is the result of bad middleware architecture mixed with an unsupported feature.

**Comment by dotnil** [comment]
If the use of `this.body` is just to passing around variables between middlewares, I'd suggest store the variable on this.state instead. `this.body` is a setter function and is expected to have side effects.

**Comment by jonathanong** [comment]
this error should be handled by these lines:

- https://github.com/koajs/koa/blob/9a9949f9eef95fbe95e11ada5b677dd7aa284fce/lib/application.js#L195-L198
- https://github.com/koajs/koa/blob/master/lib/response.js#L136

possible ways to contribute are:

- find a test case we already have for this. if we have it, we can close this issue!
- make a PR for the failing test case
- fix the test case
- do the above tasks for both v1 and v2

Note that `.respond = false` is a second-class citizen and should be used sparingly

**Comment by ryzokuken** [comment]
@jonathanong I would love to pick this up as my first PR on this project. I would need some help getting acquainted though. Firstly, do I need to make two PRs for both `master` branch and `v1.x` branch?

**Comment by jonathanong** [comment]
@ryzokuken sorry for the late reply. yes, there should be two PRs

**Comment by jonathanong** [comment]
@ryzokuken you could make the master version fo https://github.com/koajs/koa/pull/1078
